### PR TITLE
Allow request hooks to update the context.

### PR DESCRIPTION
**Pull request by wcn3**
It's useful for a request hook to enrich the context with information,
so it can pass values to downstream processing. The hook should return the
originally provided context if no modifications are made.

### CODE REVIEW

**Comment by wcn3** [comment]
R: @herohde @lostluck

File: sdks/go/pkg/beam/core/util/hooks/hooks.go [hunk 1]
```diff
@@ -87,14 +87,14 @@ func RunInitHooks(ctx context.Context) error {
...
-type RequestHook func(context.Context, *fnpb.InstructionRequest) error
+type RequestHook func(context.Context, *fnpb.InstructionRequest) (context.Context, error)
 
 // RunRequestHooks runs the hooks that handle a FnAPI request.
 func RunRequestHooks(ctx context.Context, req *fnpb.InstructionRequest) {
 	// The request hooks should not modify the request.
 	for n, h := range activeHooks {
 		if h.Req != nil {
-			if err := h.Req(ctx, req); err != nil {
+			if ctx, err := h.Req(ctx, req); err != nil {
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/util/hooks/hooks.go]: Should the same context be threaded through each hook, and then the final context returned from RunRequestHooks? At present the new contexts are all thrown away.

Comment (wcn3) [reply sdks/go/pkg/beam/core/util/hooks/hooks.go]: Fixed.

File: sdks/go/pkg/beam/core/runtime/harness/harness.go [hunk 2]
```diff
@@ -109,7 +109,7 @@ func Main(ctx context.Context, loggingEndpoint, controlEndpoint string) error {
 			log.Debugf(ctx, "RECV: %v", proto.MarshalTextString(req))
 			recordInstructionRequest(req)
 
-			hooks.RunRequestHooks(ctx, req)
+			ctx = hooks.RunRequestHooks(ctx, req)
```
Comment (herohde) [inline sdks/go/pkg/beam/core/runtime/harness/harness.go]: We should use a local variable for the returned context to limit the scope to just the present request/response.

File: sdks/go/pkg/beam/core/util/hooks/hooks.go (hunk 1 above)
Comment (herohde) [reply sdks/go/pkg/beam/core/util/hooks/hooks.go]: This function seems to still return the original context due to "ctx" defined here shadowing, not overwriting, the argument. Maybe add a test?

File: sdks/go/pkg/beam/core/runtime/harness/harness.go (hunk 2 above)
Comment (wcn3) [reply sdks/go/pkg/beam/core/runtime/harness/harness.go]: The ctx is now passed into the closure so its lifetime is properly scoped.

File: sdks/go/pkg/beam/core/util/hooks/hooks.go (hunk 1 above)
Comment (wcn3) [reply sdks/go/pkg/beam/core/util/hooks/hooks.go]: Yep. Added test. Also added context propagation to init so those hooks can establish the base context for the harness.

**Pull request merged**